llvm/test/CodeGen/PowerPC/pr40922.ll

```diff
@@ -23,11 +23,12 @@ define i32 @a() {
 ; CHECK-NEXT:    li 5, 0
 ; CHECK-NEXT:    mr 30, 3
 ; CHECK-NEXT:    addic 6, 4, 6
-; CHECK-NEXT:    addze 5, 5
 ; CHECK-NEXT:    rlwinm 6, 6, 0, 28, 26
-; CHECK-NEXT:    andi. 5, 5, 1
+; CHECK-NEXT:    addze 5, 5
 ; CHECK-NEXT:    cmplw 1, 6, 4
-; CHECK-NEXT:    crorc 20, 1, 4
+; CHECK-NEXT:    andi. 5, 5, 1
+; CHECK-NEXT:    crnot 20, 4
+; CHECK-NEXT:    cror 20, 1, 20
```

Comment:
This becomes worse. A combine for i1 type seems broken. Let me check why this happens.

Reply:
```
  t43: i32,ch = load<(dereferenceable load (s32) from @a.b)> t10, t63, undef:i32
  t44: i32,glue = addc t43, Constant:i32<6>
            t46: i32 = and t44, Constant:i32<-17>
          t48: i1 = setcc t46, t43, setuge:ch
        t54: i1 = freeze t48    ;;This freeze breaks the instruction selection pattern for or(setcc) in PPC td files.
      t55: i1 = or t58, t54
```

Reply:
I think this is a situation when we do not fold the freeze over an operation (in this case a setcc) due to the operation having more then one operand that may be poison.
However, we got
```
    t43: i32,ch = load<(dereferenceable load (s32) from @a.b)> t10, GlobalAddress:i32<ptr @a.b> 0, undef:i32
    t44: i32,glue = addc t43, Constant:i32<6>
    t46: i32 = and t44, Constant:i32<-17>
    t48: i1 = setcc t46, t43, setuge:ch
    t54: i1 = freeze t48
```
so I think that t46 and t43 only could be poison/undef if t43 is poison/undef. So it would be enough to freeze t43, like this:
```
    t43: i32,ch = load<(dereferenceable load (s32) from @a.b)> t10, GlobalAddress:i32<ptr @a.b> 0, undef:i32
    t99: i32 = freeze t43
    t44: i32,glue = addc t99, Constant:i32<6>
    t46: i32 = and t44, Constant:i32<-17>
    t54: i1 = setcc t46, t99, setuge:ch
```

Not exactly sure, but maybe if it would be possible to improve the analysis in DAGCombiner::visitFREEZE. Idea would be to allow pushing freeze through more operations (with multiple maybe poison/undef operands as we do for BUILD_VECTOR etc)). One could limit it to the scenraio when the isGuaranteedNotToBeUndefOrPoison checks for those operands reaches the same DAG node when recursing, i.e. when the likelyhood is that we still would end up with a single freeze in the end.